### PR TITLE
PT100 sensor type

### PR DESCRIPTION
**Pull request by arhi**
- added a new sensor type PT100
- added PT100 because there's a smoothieboard compatible "CPU replacement" for UP! printers that runs smoothieware and UP! printers use PT100 for both bed and hotend
- PT100 curve is different from standard NTC curve used in Thermistor sensor type hence I feel it's cleaner to add PT100 as separate sensor type then to force fit the PT100 into Thermistor class

example config:

```
temperature_control.hotend.enable              true
temperature_control.hotend.sensor              pt100
temperature_control.hotend.thermistor_pin      0.23
temperature_control.hotend.ampmod1_pin         1.20           # UP! printer uses this to "energize" the RTD
temperature_control.hotend.ampmod2_pin         1.21           # set as nc if you don't need to energize RTD
temperature_control.hotend.slope               0.0257604875
temperature_control.hotend.yintercept         -18.54
temperature_control.hotend.heater_pin          2.7
temperature_control.hotend.set_m_code          104
temperature_control.hotend.set_and_wait_m_code 109
temperature_control.hotend.designator          T
temperature_control.hotend.max_temp            300
temperature_control.hotend.min_temp            5

temperature_control.bed.enable               true
temperature_control.bed.sensor               pt100
temperature_control.bed.slope                0.0257604875
temperature_control.bed.yintercept          -18.54
temperature_control.bed.thermistor_pin       0.24
temperature_control.bed.heater_pin           2.5
temperature_control.bed.set_m_code           140
temperature_control.bed.set_and_wait_m_code  190
temperature_control.bed.designator           B
```

### CODE REVIEW

**Review by wolfmanjm — CHANGES_REQUESTED** [review]
There are superfluous includes in the pt100.h header.
Does this support other PT100 boards (like the e3d one) or is it UP only?

**Comment by arhi** [comment]
includes, sorry I forgot about those sorry :D .. removing asap
other boards - I tested only UP!Plus2 but looking at the 
http://wiki.e3d-online.com/wiki/E3D_PT100_Amplifier_Documentation
that one should work too without a problem, it only needs different config

```
temperature_control.hotend.enable              true
temperature_control.hotend.sensor              pt100
#temperature_control.hotend.ampmod1_pin         nc
#temperature_control.hotend.ampmod2_pin         nc
temperature_control.hotend.slope               0.3173828125
temperature_control.hotend.yintercept         -299.6
```

by my calc that's +2 / -1 C error on the 160-280C span but I might be wrong, can't be 100% sure without trying the board myself but it's just a matter of setting a proper slope and yintercept parameters

**Comment by arhi** [comment]
@ wolfmanjm removed superfluous includes in the pt100.h and pt100.cpp

**Comment by arhi** [comment]
I have zero experience with "rake" so not sure about smoothieware make procedure .. I added in the code the #ifndef NO_RTD_CURVE not to waste flash with fancy math if one don't want it .. but I have no clue how to add this -DNO_RTD_CURVE to the build system except manually editing makefiles ?! ... I was looking at the NONETWORK but looks like again I need to manually edit makefile ?

File: ConfigSamples/UP2Plus_tinyfabCPU/config [hunk 1]
```diff
@@ -0,0 +1,219 @@
+# NOTE Lines must not exceed 132 characters
```
Comment (wolfmanjm) [inline ConfigSamples/UP2Plus_tinyfabCPU/config]: remove this file we do not allow samples for proprietary machines.
You can add a snippet for the PT100 though.

**Review by wolfmanjm — CHANGES_REQUESTED** [review]

**Comment by wolfmanjm** [comment]
Please remove the #ifndefs we do not allow conditional compilation in smoothie for submitted things like this.

**Comment by arhi** [comment]
no problem removing asap didn't know that

**Review by wolfmanjm — APPROVED** [review]
Thanks. Smoothie is designed to be a binary distribution 100% configured by the text config.

**Comment by arhi** [comment]
@wolfmanjm just trying to be helpful :D

> 100% configured by the text config.

On that note I just made some calculations and I see that any AMP configuration I look at ends up being in same format after resolving it to simpler form (x * adc)/(y-z*adc) .. so I'll just add these xyz values to be configurable from text config too :D making this more versatile

**Comment by wolfmanjm** [comment]
good idea, just use good defaults.

**Comment by arhi** [comment]
I have a fairly useful .ods for visualising and testing all the values
.. dunno if there's a proper place in the code to add it or it's
something for forum/website ?

**Comment by wolfmanjm** [comment]
put it on the wiki under temperature control in a section for PT100. - Thanks

**Comment by arhi** [comment]
@wolfmanjm I didn't notice that whole site is a wiki :D ... done
wrote some story about PT100, some formulas .. I hope that's ok and uploaded file
http://smoothieware.org/3d-printer-guide#toc12
http://smoothieware.org/local--files/3d-printer-guide/PT100%20E3D.ods

**Comment by wolfmanjm** [comment]
That is great except you put it in the wrong place, it needs to go in this page http://smoothieware.org/temperaturecontrol Just cut it and paste it in the correct place. - Thanks

**Comment by arhi** [comment]
@wolfmanjm hopefully it's ok now

**Comment by arhi** [comment]
Hi @wolfmanjm I'm running this for a while now without any issues on two machines, works like a charm and here: https://github.com/Smoothieware/Smoothieware/issues/683#issuecomment-257386036 the @pyjamasam tested it with his e3d amp and he's in with "within decimals on every temp I have tried." so dunno if any more testing is required, it does work nicely :D ..
 so if there's anything else I should do to make this mergeable lemme know :)

**Comment by arthurwolf** [comment]
What is missing for this to be merged at this point ?

**Comment by arthurwolf** [comment]
The current documentation at http://smoothieware.org/pt100 is *way* too technical and it's nearly impossible for a newcomer to extract information about how to use a PT100 from it. Would it be possible to rewrite it as a much shorter and much simpler how-to guide ? I started it at the top of the page.
Thanks !

**Comment by wolfmanjm** [comment]
You will need to redo this pr you ignored the instructions in the contribution guidelines and this can no longer be merged.

**Comment by arhi** [comment]
didn't ignore nothing, sorry, I tried to update my fork with latest work you guy's did to update my printer but somehow messed it up and merged master in to edge or something like that ... so after trying to revert it back I just deleted it, safest way to solve the problem without messing something else up. Sorry if that introduced any issues to you guy's.

**Comment by wolfmanjm** [comment]
I don't know what you did (looks like you merged master into edge) but the git repo for this is now totally foobar, and I do not know how it can be recovered, so you will need to start again from edge, and ONLY add the pt100 related patches.
You could try reverting the last set of changes but I doubt that will fix it. You need to be much more careful how you use git in future, if you are unsure ask before you push.

**Comment by arhi** [comment]
@wolfmanjm exactly what I did :( I tried to fetch the latest changes from your repo so that I have latest smoothie on my printer with pt100 and then somehow I messed it up 
(what I exactly did is "git pull upstream master" instead of "git pull upstream edge" I think)
anyhow, deleted the repo already as after trying for a while to revert the messup I gave up.
sorry for any inconvenience. I made a new fork and will insert the pt100 stuff in it in few minutes

**Comment by arhi** [comment]
created PR#1086

**Comment by DeathmannX** [comment]
Hi Guys, i tryed to set up on my Printer the PT100 amplifyed with the Kit from e3D but it displays only:

> ok T0:inf/0.0 @0 B21.8/0.0 @0

So the Heatbed Thermistor can be get but the pt100still hates me. 😞 
What do i need to add in my config?

**Comment by arhi** [comment]
what do you have in your config now?
something like this should work:


```
temperature_control.hotend.thermistor_pin 0.23
temperature_control.hotend.pt100linear    0
temperature_control.hotend.pt100_a        0.0039083
temperature_control.hotend.pt100_b        0.0000005775
temperature_control.hotend.pt100_r0       100
temperature_control.hotend.pt100_amptype  0

temperature_control.hotend.pt100_q1  810000
temperature_control.hotend.pt100_q2  45409000
temperature_control.hotend.pt100_q3  280

# quoeficients that are also known to work with E3D amp:
#temperature_control.hotend.pt100_q1            11000000
#temperature_control.hotend.pt100_q2            606888919
#temperature_control.hotend.pt100_q3            2500


```

you can tweak q1/q2/q3 depending on how you are powering the amp and of course setup your own temperature_control.hotend.thermistor_pin and make sure you connect the amp output to ADC pin that does not have 4k7 pullup resistor

**Comment by DeathmannX** [comment]
@arhi This is what i got:


(better to read) ;)
http://pastebin.com/12GMcGU9

**Comment by arhi** [comment]
that sounds about right ... it say that it's infinite, that is basically error reported .. what voltage you have on the 0.27 pin?

**Comment by DeathmannX** [comment]
ok got the Problem...
I tryed to use a 3.3V pin instat a 5V pin

**Comment by arhi** [comment]
ah yes, those q1/q2/q3 are for powering the E3D amp from 5V ... but it is **UNSAFE**!! as it can have then on output 5V and that will kill the ADC pin (and maybe even the whole mcu) .. 

using a 3v0 zenner clamp on the ADC imput is required if you are powering E3D amp from 5V

**Comment by arhi** [comment]
if you power the E3D amp from 3v3 these Q1/Q2/Q3 should work ok:

```
temperature_control.hotend.pt100_q1 = 4400
temperature_control.hotend.pt100_q2 = 163770
temperature_control.hotend.pt100_q3 = 1
```

**Comment by DeathmannX** [comment]
nope, this Values didn't work.

It only shows me:
`ok T0:inf/0.0 @0 B21.8/0.0 @0`

**Comment by arhi** [comment]
when you power the e3d board from 3v3 what's the voltage on the adc pin of the smoothiebord?

**Comment by DeathmannX** [comment]
The + get 3.36V and to the Board goes 0.80V

**Comment by arhi** [comment]
>
> The + get 3.36V and to the Board goes 0.80V
>

that should give ADC reading of 3970.909090909091
which is equivalent to 109.34 OHM or 23.88C

that's assuming you are using Q1 = 4400; Q2 = 163770 and Q3 = 1

lemme go trough the source quickly..

infinity is returned:


if(bad_config) return infinityf();
...
    const uint32_t max_adc_value= THEKERNEL->adc->get_max_value();
    if ((adc_value >= max_adc_value) || (adc_value == 0))
        return infinityf();
...
    if (this->use_linear){
...
    } else {
...
        switch (this->amptype){
...
        default: // TODO: add more amp types
           t = infinityf();
```

so i don't see it returning inf except for bad config :( ...

**Comment by arhi** [comment]
you have this in your config:

`temperature_control.hotend.thermistor           pt100
`

that is wrong :( this is not a thermistor :) .. you need to define pt100 sensor type, not to setup pt100 thermiostor type  ... so use temperature_control.hotend.sensor. without it you are not loading pt100 object at all

full temp setup for pt100:

```
## Temperature control configuration
temperature_control.hotend.enable              true
temperature_control.hotend.sensor              pt100     #THIS IS WHAT YOU ARE MISSING!
temperature_control.hotend.thermistor_pin      0.23
temperature_control.hotend.set_m_code          104
temperature_control.hotend.set_and_wait_m_code 109
temperature_control.hotend.designator          T
temperature_control.hotend.max_temp            400
temperature_control.hotend.min_temp            5
temperature_control.hotend.p_factor            23.0
temperature_control.hotend.i_factor            1.104
temperature_control.hotend.d_factor            120
temperature_control.hotend.pt100_q1   4400
temperature_control.hotend.pt100_q2   163770
temperature_control.hotend.pt100_q3   1

temperature_control.bed.enable                 true
temperature_control.bed.sensor                 pt100
temperature_control.bed.slope                  0.0234092253
temperature_control.bed.yintercept            -2.85
temperature_control.bed.thermistor_pin         0.24
temperature_control.bed.heater_pin             2.5
temperature_control.bed.set_m_code             140
temperature_control.bed.set_and_wait_m_code    190
temperature_control.bed.designator             B
temperature_control.bed.bang_bang              true
temperature_control.bed.hysteresis             1.0
temperature_control.bed.min_temp               5
temperature_control.bed.max_temp               200

```

**Comment by DeathmannX** [comment]
Ok Thx now it displays not more inf.
Now the Temparture is -1.0 😄 

Maybe 3.36V make the Sensor inaccurate?
https://gyazo.com/966157aec84c25325d3e9b90c216d648

PS: The bed parth i didn't included because i got only a Hotend pt100. ;)

**Comment by DeathmannX** [comment]
ouh, maybe i did find out why i get -1

I see 2 different versions of smoothieware:
https://github.com/Smoothieware/Smoothieware
https://github.com/arhi/Smoothieware

But i see in the commits that you and wolfmanjm sending updates on both reposities.
What is the difference betwen both. (not just the pt100)?

**Comment by arhi** [comment]
the first one is original smoothieware where the pt100 pull request is not yet merged so there's no pt100 there and you get -1 'cause pt100 sensor don't exist

second one (/arhi/Smoothieware) is my fork with the pt100 sensor

the differences are - my fork has the pt100 sensor, original smoothieware had few (82 attm) commits after the fork I made, nothing relevant for the testing of pt100 sensor... number of improvements of course so you want to finally go with original one but for testing pt100 please use https://github.com/arhi/Smoothieware

now I should be able to update my fork with all of these 82 commits smothieware team made but last time I did that I had to create everything from scratch 'cause it got messed up so attm I'm just leaving it as is for the team to merge or not the changes... the fork is safe (won't break anything) and works for a long time on few of my printers, so it's safe for testing pt100 sensor :)

thanks for testing btw
b.

**Comment by DeathmannX** [comment]
cool sounds good. ;)
I will give you as mutch feedback as i can if i'am done with the batchfile of you're Smoothieversion. 
https://gyazo.com/b35ac78c7fd511e172d0694eef4f4783

**Comment by DeathmannX** [comment]
tryed it with you're modifyed Smoothieversion and it didn't worked for me.

It displays only -1 for the Hotend.

**Comment by arhi** [comment]
something is wrong, either the compiled binary or the config as that -1 is error reported by smoothieware .. here's bin and config that works, remove bed part of the config completely for test, change only the pin of the adc pin for start and test like that, temp will be off but you should see something that's not -1 nor inf, then if that works change/add 

```
temperature_control.hotend.pt100linear    0
temperature_control.hotend.pt100_amptype  0
temperature_control.hotend.pt100_q1   4400
temperature_control.hotend.pt100_q2   163770
temperature_control.hotend.pt100_q3   1

```
here's the binary
http://wikisend.com/download/341666/SMOTHIE.zip

**Comment by DeathmannX** [comment]
same Result :(

**Comment by arhi** [comment]
hm I'll check it on smoothieboard tomorrow afternoon ... send me your config

**Comment by DeathmannX** [comment]
My normal config where my Extruder an other stuff is configurated for my Re-ARM + Ramps 1.4 Board.
https://www.file-upload.net/download-12346163/config.txt.html

The config where i changed just the pin for my pt100
https://www.file-upload.net/download-12346169/config.txt.html

**Comment by arhi** [comment]
that site want me to get their download.exe .. that's not happening ... 

copy/paste your config here: http://pastebin.com/

**Comment by DeathmannX** [comment]
sry, i didn't know that. 
http://pastebin.com/mU6qQpE2

http://pastebin.com/f2G3W81p

**Comment by arhi** [comment]
@DeathmannX  your E3D AMP PT100 is connected to 0.27 pin right?

**Comment by arhi** [comment]
@DeathmannX !?
0.27 is P0[27]/SDA0/USB_SDA... I don't see it can be ADC ?

in your orig. u use 0.25, that's P0[25]/AD0[2]/I2SRX_SDA/TX so can be ADC (like 0.24 and 0.26) but 0.27 from what I see can't be ...

can you use any of the ADC pins? 0.23, 0.24, 0.25, 0.26, 0.2, 0.3, 1.30, 1.31
0.27 won't work so I guess that's where the -1 comes from.

**Comment by DeathmannX** [comment]
This is my Board:
https://www.kickstarter.com/projects/1245051645/re-arm-for-ramps-simple-32-bit-upgrade
The Ramps 1.4 goes on top.

Here is the pinout:
http://bit.ly/2mdHZ3S

Here is where are the pins on the Ramps 1.4 Board:
http://wiki.e3d-online.com/images/thumb/4/4b/RAMPSPT100.jpg/800px-RAMPSPT100.jpg

I tryed every possible pin and get now 5V ;)
But -1 didn't changed at 1 of them.
And yes i changed the q1, q2 and q3 back to 

`temperature_control.hotend.pt100_q1  810000
 temperature_control.hotend.pt100_q2  45409000
 temperature_control.hotend.pt100_q3  280`

So it can only be something in the code i belive.

**Comment by arhi** [comment]
pinout link don't work :( (never seen that board before)
EDIT: found the pinout on S3.. will look in a sec

the PT100 module can only return values between 5 and infinite, -1 is returned by the smoothieware before pt100 module is loaded. so if you are getting -1 something else is wrong.

are you sure you changed the firmware on the board and put one I sent? 'cause I loaded that firmware on my dev smoothie capable board with your config and it works with your config

**Comment by DeathmannX** [comment]
Pinout re upload (pdf):
https://drive.google.com/file/d/0BxLYV9G8oHT2eV9ENGo1Vi16Z1E/view?usp=sharing

Yes i'am 100% shure cause my Display didn't work with the config and the firmware. ;)

**Comment by arhi** [comment]
any LED linked on some pin that's not used?

**Comment by DeathmannX** [comment]
no

**Comment by arhi** [comment]
I see a "play" led, does it exist on a board or only on ramps?
will try to use that one ... if it is not on the board, maybe you have a led and res you can connect to pin13 :)

here's the config you can try, don't need to connect ramps board on the rearm at all, just let rearm board run (if that's possible, where's the sd card? on rearm itself or ?
```

extruder.hotend.enable                       true
extruder.hotend.step_pin                     2.0           
extruder.hotend.dir_pin                      0.5          
extruder.hotend.en_pin                       0.4          

temperature_control.hotend.enable              true
temperature_control.hotend.sensor              pt100
temperature_control.hotend.thermistor_pin      0.23  #T0/HE1
temperature_control.hotend.heater_pin          2.5
temperature_control.hotend.ampmod1_pin         4.28
temperature_control.hotend.ampmod2_pin         nc
temperature_control.hotend.set_m_code          104
temperature_control.hotend.set_and_wait_m_code 109
temperature_control.hotend.designator          T
temperature_control.hotend.max_temp            400
temperature_control.hotend.min_temp            5
temperature_control.hotend.p_factor            23.0
temperature_control.hotend.i_factor            1.104
temperature_control.hotend.d_factor            120
temperature_control.hotend.pt100linear    0
temperature_control.hotend.pt100_amptype  0
temperature_control.hotend.pt100_q1   4400
temperature_control.hotend.pt100_q2   163770
temperature_control.hotend.pt100_q3   1

panel.enable                                 false
network.enable                               false
laser_module_enable                          false



```

**Comment by pyjamasam** [comment]
From those pinouts images and the documentation for that board it looks like A9 is mapped to 0.26
A9 seems to be in AUX-2 on the ramps board.

**Comment by DeathmannX** [comment]
> From those pinouts images and the documentation for that board it looks like A9 is mapped to 0.26
A9 seems to be in AUX-2 on the ramps board.
Thats right.

@arhi sry, i got no led. :(

**Comment by arhi** [comment]
```

extruder.hotend.enable                       true
extruder.hotend.step_pin                     2.0           
extruder.hotend.dir_pin                      0.5          
extruder.hotend.en_pin                       0.4          

temperature_control.hotend.enable              true
temperature_control.hotend.sensor              pt100
temperature_control.hotend.thermistor_pin      0.23  #T0/HE1
temperature_control.hotend.heater_pin          2.5
temperature_control.hotend.ampmod1_pin         4.28
temperature_control.hotend.ampmod2_pin         nc
temperature_control.hotend.set_m_code          104
temperature_control.hotend.set_and_wait_m_code 109
temperature_control.hotend.designator          T
temperature_control.hotend.max_temp            400
temperature_control.hotend.min_temp            5
temperature_control.hotend.p_factor            23.0
temperature_control.hotend.i_factor            1.104
temperature_control.hotend.d_factor            120
temperature_control.hotend.pt100linear    0
temperature_control.hotend.pt100_amptype  0
temperature_control.hotend.pt100_q1   4400
temperature_control.hotend.pt100_q2   163770
temperature_control.hotend.pt100_q3   1

panel.enable                                 false
network.enable                               false
laser_module_enable                          false

```

nothing else in the config.txt - just these lines.. 

login to smoothieware using terminal and enter "get temp" and press enter
you need to get some random value (around 180C I suspect will be the floating voltage) ... so if SD card is on that board you don't even connect your ramps board on the rearm... 

once you get that working, you can start adding things to your config and finally returning back your ramps board

**Comment by arhi** [comment]
image

**Comment by arhi** [comment]
also with the config I shown here on the pin 4.28 (pin13 on that arduino connector where normally led goes) every time temperature is read that pin will "flash" (it goes on when the temp is read and then back off after adc reading is done) so if you attach a led to it (usually arduino has a led on pin13), or a scope, or analog voltmeter you will see if the pt100 module is reading temp at all

can't really go farther then this :( .. as I said, -1 is returned by smoothieware if something is wrong, pt100 returns either the temp value (has to be greater then 5) or infinite...

**Comment by DeathmannX** [comment]
This RealTerm Software displays only a black therminal.

EDITE:
After restarting the tool it says Smoothie ok
but where can i say get temp

**Comment by DeathmannX** [comment]
`T <57988> temp: -1.00000/0.00000 @0`

**Comment by arhi** [comment]
> T <57988> temp: -1.00000/0.00000 @0

with *only* those 28 lines in config?
with nothing but your re-arm board connected to usb with sd card in it, no ramps board on top, nothing on top, plain re-arm on usb ?

what does these return?

```
config-get temperature_control.hotend.sensor
config-get temperature_control.hotend.pt100linear
config-get temperature_control.hotend.pt100_amptype

```
image

really have no clue what is wrong, that firmware that I sent you, with this config that I pasted few posts up (so nothing in but hotend, 28 lines total), pushed onto nxp mcu works.. I know nxp arm's have bugs in them but not these type of bugs..

**Comment by DeathmannX** [comment]
I got something:
It displays now 135

http://pastebin.com/29qNfSTy

I can use you're firmware or the normal it dosen't matter.

**Comment by arhi** [comment]
that config is not using pt100 sensor type. 
just use the 28 line config I provided

**Comment by DeathmannX** [comment]
I used a other Semitec Sensor with a few Pins i tested with the pt100 and that worked.
So my Pins do they Job just fine. ;)

**Comment by arhi** [comment]
dunno 'bout "your pins", as I said, that config you send is not using pt100 module, if you don't configure pt100 module -> you are not testing pt100 module...

the 28line config I sent you uses pt100 module and can help you debug what in your config is wrong so you get -1... far from that I can't help (I have here 2 smoothieboard compatible board that work, and also @pyjamasam tested it as you can see [here](https://github.com/Smoothieware/Smoothieware/issues/683) with his smoothieboard and it works for him too.. also as I said the pt100 module will never return -1, -1 is returned by smoothieware when something is wrong)

**Comment by DeathmannX** [comment]
I tested your files with the same result. - 1
So I tested that but I didn't noticed that I used the wrong sensor. :D

**Comment by arhi** [comment]
As I already write, use the config-get to see if you are using the proper sensor module

```
config-get temperature_control.hotend.sensor
config-get temperature_control.hotend.pt100linear
config-get temperature_control.hotend.pt100_amptype
```

**Pull request closed without merging**